packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,514 @@
+    const parentThreadId = outOfBandMessage.id
+
+    const message = new DidExchangeRequestMessage({ label, parentThreadId, did: peerDid.did, goal, goalCode })
+
+    // Create sign attachment containing didDoc
+    if (peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc) {
+      const didDocAttach = await this.createSignedAttachment(didDocument, [verkey].map(didKeyToVerkey))
+      message.didDoc = didDocAttach
+    }
+
+    connectionRecord.did = peerDid.did
+    connectionRecord.threadId = message.id
+
+    if (autoAcceptConnection !== undefined || autoAcceptConnection !== null) {
+      connectionRecord.autoAcceptConnection = autoAcceptConnection
+    }
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Create message ${DidExchangeRequestMessage.type} end`, {
+      connectionRecord,
+      message,
+    })
+    return { message, connectionRecord }
+  }
+
+  public async processRequest(
+    messageContext: InboundMessageContext<DidExchangeRequestMessage>,
+    outOfBandRecord: OutOfBandRecord,
+    routing?: Routing
+  ): Promise<ConnectionRecord> {
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} start`, messageContext)
+
+    // TODO check oob role is sender
+    // TODO check oob state is await-respons
+    // TODO check there is no connection record for particular oob record
+
+    const { did, mediatorId } = routing ? routing : outOfBandRecord
+    if (!did) {
+      throw new AriesFrameworkError('Out-of-band record does not have did attribute.')
+    }
+
+    const { message } = messageContext
+
+    // Check corresponding invitation ID is the request's ~thread.pthid
+    // TODO Maybe we can do it in handler, but that actually does not make sense because we try to find oob by parent thread ID there.
+    if (!message.thread?.parentThreadId || message.thread?.parentThreadId !== outOfBandRecord.getTags().messageId) {
+      throw new DidExchangeProblemReportError('Missing reference to invitation.', {
+        problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+      })
+    }
+
+    // If the responder wishes to continue the exchange, they will persist the received information in their wallet.
+
+    if (!message.did.startsWith('did:peer:')) {
+      throw new DidExchangeProblemReportError(
+        `Message contains unsupported did ${message.did}. Supported dids are [did:peer]`,
+        {
+          problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+        }
+      )
+    }
+    const peerDid = DidPeer.fromDid(message.did)
+    if (peerDid.numAlgo !== PeerDidNumAlgo.GenesisDoc) {
+      throw new DidExchangeProblemReportError(
+        `Unsupported numalgo ${peerDid.numAlgo}. Supported numalgos are [${PeerDidNumAlgo.GenesisDoc}]`,
+        {
+          problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+        }
+      )
+    }
```

Comment:
I think I was because of this https://github.com/hyperledger/aries-rfcs/issues/717. As reading the related PR https://github.com/hyperledger/aries-rfcs/pull/725, I don't understand why the request doesn't have to be signed, but the response does.

Also, would that mean that numAlgo 2 will be supported for request and not for response?

Reply:
The response only needs to be signed when the invitation key is different from the key that will be used for the connection. As the request doesn't have an invitation it will never need to sign something. The RFC was incorrect in that both should contain an signature.

But good point on what we should sign then. I added a comment to the PR